File: web/src/types/resume.ts
/**
 * 全局类型定义
 * 与后端 app/models/resume.py 中的 Pydantic 模型一一对应。
 * 所有可选字段均为 null（而非 undefined），与后端 JSON 序列化行为一致。
 */

/** 候选人基本联系信息 */
export interface BasicInfo {
  name: string | null;
  phone: string | null;
  email: string | null;
  address: string | null;
  gender: string | null;
  birth_date: string | null;
  github: string | null;
  linkedin: string | null;
  website: string | null;
  wechat: string | null;
}

/** 候选人求职意向信息 */
export interface JobInfo {
  intention: string | null;       // 目标岗位
  expected_salary: string | null;
  job_type: string | null;        // 全职/兼职/实习
  available_date: string | null;  // 到岗时间
  preferred_location: string | null;
}

/** 单条教育经历 */
export interface EducationItem {
  school: string | null;
  degree: string | null;      // 学历层次：本科/硕士/博士/专科
  major: string | null;
  start_date: string | null;
  end_date: string | null;
  gpa: string | null;
  description: string | null; // 在校经历、荣誉、社团等
}

/** 单条工作经历 */
export interface WorkExperienceItem {
  company: string | null;
  position: string | null;
  start_date: string | null;
  end_date: string | null;
  location: string | null;
  description: string | null;
}

/** 单条项目经历 */
export interface ProjectItem {
  name: string | null;
  role: string | null;
  start_date: string | null;
  end_date: string | null;
  tech_stack: string[];       // 技术栈列表，每项为独立技术名称
  description: string | null;
  url: string | null;         // 项目链接或 GitHub 地址
}

/** 候选人背景信息（教育、工作、项目经历及技能） */
export interface BackgroundInfo {
  years_of_experience: string | null;
  education: string | null;              // 最高学历（简要，如"本科"）
  education_list: EducationItem[];
  work_experience: WorkExperienceItem[];
  project_experience: ProjectItem[];
  skills: string[];
  certifications: string[];
  languages: string[];
  awards: string[];
  publications: string[];
  open_source: string[];
}

/** 完整的简历信息（LLM 提取结果） */
export interface ResumeInfo {
  basic_info: BasicInfo;
  job_info: JobInfo;
  background_info: BackgroundInfo;
  raw_text: string;          // PDF 原始文本，供评分时作为后备参考
  enriched_context: string;  // 从外部链接抓取的补充信息
}

/** Redis 中存储的完整简历解析结果 */
export interface ResumeAnalysisResponse {
  resume_id: string;
  resume_info: ResumeInfo;
  cached: boolean;
}

/** 上传简历接口的响应 */
export interface ResumeSubmitResponse {
  resume_id: string;
}

/** 简历解析状态 */
export type ResumeParseStatus = "pending" | "done" | "error";

/** 查询简历解析状态的响应 */
export interface ResumeStatusResponse {
  resume_id: string;
  status: ResumeParseStatus;
  resume_info: ResumeInfo | null;
  cached: boolean;   // true 表示命中缓存，未重新调用 LLM
  error: string | null;
}

/** 简历与岗位的匹配评分结果 */
export interface MatchResult {
  score: number;                  // 综合评分，0-100
  skill_match_rate: number;       // 技能匹配率，0-1
  experience_relevance: number;   // 经验相关性，0-1
  tolerance_score: number;        // 适应潜力，0-1（考虑迁移学习能力）
  analysis: string;               // 综合评估说明
  growth_outlook: string;         // 入职后成长走向预测
  matched_keywords: string[];     // 已匹配关键词（含推断出的隐含技能）
  missing_keywords: string[];     // 缺失关键词（整个领域无经验时才列出）
  transferable_skills: string[];  // 可迁移技能（有相关经验但非直接匹配）
}

/** 同步匹配接口的响应（当前未使用，保留备用） */
export interface MatchResponse {
  resume_id: string;
  job_description: string;
  match_result: MatchResult;
  cached: boolean;
}

/** 匹配任务状态 */
export type MatchJobStatus = "pending" | "done" | "error";

/** 提交匹配任务接口的响应 */
export interface MatchSubmitResponse {
  match_id: string;
}

/** 查询匹配任务状态的响应 */
export interface MatchStatusResponse {
  match_id: string;
  status: MatchJobStatus;
  match_result: MatchResult | null;
  cached: boolean;
  error: string | null;
}

File: app/models/resume.py
from pydantic import BaseModel


class BasicInfo(BaseModel):
    name: str | None = None
    phone: str | None = None
    email: str | None = None
    address: str | None = None
    gender: str | None = None
    birth_date: str | None = None
    github: str | None = None
    linkedin: str | None = None
    website: str | None = None
    wechat: str | None = None


class EducationItem(BaseModel):
    school: str | None = None
    degree: str | None = None  # 学历：本科/硕士/博士等
    major: str | None = None
    start_date: str | None = None
    end_date: str | None = None
    gpa: str | None = None
    description: str | None = None  # 在校经历、荣誉等


class WorkExperienceItem(BaseModel):
    company: str | None = None
    position: str | None = None
    start_date: str | None = None
    end_date: str | None = None
    location: str | None = None
    description: str | None = None  # 工作内容、成就


class ProjectItem(BaseModel):
    name: str | None = None
    role: str | None = None
    start_date: str | None = None
    end_date: str | None = None
    tech_stack: list[str] = []
    description: str | None = None
    url: str | None = None  # 项目链接/GitHub


class JobInfo(BaseModel):
    intention: str | None = None
    expected_salary: str | None = None
    job_type: str | None = None  # 全职/兼职/实习
    available_date: str | None = None  # 到岗时间
    preferred_location: str | None = None


class BackgroundInfo(BaseModel):
    years_of_experience: str | None = None
    education: str | None = None  # 最高学历（简要）
    education_list: list[EducationItem] = []  # 完整教育经历
    work_experience: list[WorkExperienceItem] = []  # 工作经历
    project_experience: list[ProjectItem] = []  # 项目经历
    skills: list[str] = []  # 技能列表
    certifications: list[str] = []  # 证书/资质
    languages: list[str] = []  # 语言能力
    awards: list[str] = []  # 获奖经历
    publications: list[str] = []  # 论文/出版物
    open_source: list[str] = []  # 开源贡献


class ResumeInfo(BaseModel):
    basic_info: BasicInfo = BasicInfo()
    job_info: JobInfo = JobInfo()
    background_info: BackgroundInfo = BackgroundInfo()
    raw_text: str = ""
    enriched_context: str = ""


class MatchResult(BaseModel):
    score: float  # 0-100
    skill_match_rate: float
    experience_relevance: float
    tolerance_score: float = 0.0  # 0-1，考虑迁移学习能力后的潜力评分
    analysis: str
    growth_outlook: str = ""  # 入职后可能的走向与成长预测
    matched_keywords: list[str] = []
    missing_keywords: list[str] = []
    transferable_skills: list[str] = []  # 可迁移技能（有相关经验但非直接匹配）


class ResumeSubmitResponse(BaseModel):
    resume_id: str


class ResumeStatusResponse(BaseModel):
    resume_id: str
    status: str  # "pending" | "done" | "error"
    resume_info: ResumeInfo | None = None
    cached: bool = False
    error: str | None = None


class ResumeAnalysisResponse(BaseModel):
    resume_id: str
    resume_info: ResumeInfo
    cached: bool = False


class MatchResponse(BaseModel):
    resume_id: str
    job_description: str
    match_result: MatchResult
    cached: bool = False


class MatchSubmitResponse(BaseModel):
    match_id: str


class MatchStatusResponse(BaseModel):
    match_id: str
    status: str  # "pending" | "done" | "error"
    match_result: MatchResult | None = None
    cached: bool = False
    error: str | None = None
